### ISSUE TITLE
Implement Complete ConsistencyCheckerAgent with Environment and RL Components

### ISSUE THREAD

**Issue by Marwan-Adawi** [Marwan-Adawi/ABET-Accreditation-Automation-System#3]
### Components Overview

1. **Environment Implementation**
2. **State Space Definition**  
3. **Action Space Definition**
4. **Reward Function**
5. **Agent Implementation**

### Code Implementation

1. First, create the environment class in `agents/environments/consistency_checker_env.py`:

```python
from typing import Dict, List, Tuple, Optional
import numpy as np
from gymnasium import Env, spaces

class ConsistencyCheckEnvironment(Env):
    """Environment for consistency checking in ABET documentation."""
    
    def __init__(self, documents: Dict):
        super().__init__()
        
        # Store documents to check
        self.documents = documents
        
        # Define observation space components
        self.num_clos = len(self.get_all_clos())
        self.num_plos = len(self.get_all_plos())
        self.num_assessment_methods = len(self.get_assessment_methods())
        
        # State space: Binary vectors for CLO-PLO mappings, assessment methods, and scores
        self.observation_space = spaces.Dict({
            'clo_plo_mapping': spaces.Box(low=0, high=1, shape=(self.num_clos, self.num_plos)),
            'assessment_methods': spaces.Box(low=0, high=1, shape=(self.num_clos, self.num_assessment_methods)),
            'scores_valid': spaces.Box(low=0, high=1, shape=(self.num_clos,)),
            'syllabus_match': spaces.Box(low=0, high=1, shape=(self.num_clos,))
        })
        
        # Action space: For each CLO, can flag issues with:
        # - CLO-PLO mapping
        # - Assessment method
        # - Score validity
        # - Syllabus consistency
        self.action_space = spaces.Dict({
            'clo_index': spaces.Discrete(self.num_clos),
            'issue_type': spaces.Discrete(4),  # 4 types of issues
            'flag': spaces.Discrete(2)  # 0: no issue, 1: flag issue
        })
        
        # Initialize state
        self.state = self._get_initial_state()
        self.inconsistencies = []
        
    def reset(self, seed=None) -> Tuple[Dict, Dict]:
        """Reset the environment to initial state."""
        super().reset(seed=seed)
        self.state = self._get_initial_state()
        self.inconsistencies = []
        return self.state, {}
        
    def step(self, action: Dict) -> Tuple[Dict, float, bool, bool, Dict]:
        """Take action in environment."""
        clo_idx = action['clo_index']
        issue_type = action['issue_type']
        flag = action['flag']
        
        # Get current state
        prev_state = self.state.copy()
        
        # Apply action
        reward = self._apply_action(clo_idx, issue_type, flag)
        
        # Check if episode is done (all consistency checks completed)
        done = len(self.inconsistencies) == self._count_actual_inconsistencies()
        
        return self.state, reward, done, False, {}
        
    def _get_initial_state(self) -> Dict:
        """Initialize state from documents."""
        clo_plo_mapping = self._extract_clo_plo_mapping()
        assessment_methods = self._extract_assessment_methods()
        scores_valid = self._check_score_validity()
        syllabus_match = self._check_syllabus_consistency()
        
        return {
            'clo_plo_mapping': clo_plo_mapping,
            'assessment_methods': assessment_methods,
            'scores_valid': scores_valid,
            'syllabus_match': syllabus_match
        }
        
    def _apply_action(self, clo_idx: int, issue_type: int, flag: int) -> float:
        """Apply action and return reward."""
        # Get actual inconsistency status
        actual_issue = self._check_actual_inconsistency(clo_idx, issue_type)
        
        # Calculate reward
        if actual_issue and flag == 1:
            # Correctly identified inconsistency
            reward = 1.0
            self.inconsistencies.append((clo_idx, issue_type))
        elif not actual_issue and flag == 0:
            # Correctly identified consistency
            reward = 0.5
        else:
            # Incorrect identification
            reward = -1.0
            
        return reward
        
    def _check_actual_inconsistency(self, clo_idx: int, issue_type: int) -> bool:
        """Check if there is actually an inconsistency."""
        if issue_type == 0:  # CLO-PLO mapping
            return not any(self.state['clo_plo_mapping'][clo_idx])
        elif issue_type == 1:  # Assessment method
            return not any(self.state['assessment_methods'][clo_idx])
        elif issue_type == 2:  # Score validity
            return not self.state['scores_valid'][clo_idx]
        else:  # Syllabus consistency
            return not self.state['syllabus_match'][clo_idx]
            
    def _count_actual_inconsistencies(self) -> int:
        """Count total number of actual inconsistencies."""
        count = 0
        for clo_idx in range(self.num_clos):
            for issue_type in range(4):
                if self._check_actual_inconsistency(clo_idx, issue_type):
                    count += 1
        return count
```

2. Now implement the ConsistencyCheckerAgent in `agents/consistency_checker_agent.py`:

```python
from typing import Dict, List, Optional, Tuple
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from collections import deque
import random

from .environments.consistency_checker_env import ConsistencyCheckEnvironment

class ConsistencyNetwork(nn.Module):
    """Neural network for consistency checking."""
    
    def __init__(self, state_size: int, action_size: int):
        super().__init__()
        self.fc1 = nn.Linear(state_size, 128)
        self.fc2 = nn.Linear(128, 64)
        self.fc3 = nn.Linear(64, action_size)
        
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = torch.relu(self.fc1(x))
        x = torch.relu(self.fc2(x))
        return self.fc3(x)

class ConsistencyCheckerAgent:
    """Agent for checking consistency in ABET documentation."""
    
    def __init__(self, env: ConsistencyCheckEnvironment, 
                 learning_rate: float = 0.001,
                 gamma: float = 0.99,
                 epsilon_start: float = 1.0,
                 epsilon_end: float = 0.01,
                 epsilon_decay: float = 0.995,
                 memory_size: int = 10000,
                 batch_size: int = 64):
        
        # Environment
        self.env = env
        
        # State and action sizes
        self.state_size = self._calculate_state_size()
        self.action_size = self._calculate_action_size()
        
        # Neural network and optimizer
        self.network = ConsistencyNetwork(self.state_size, self.action_size)
        self.target_network = ConsistencyNetwork(self.state_size, self.action_size)
        self.target_network.load_state_dict(self.network.state_dict())
        self.optimizer = optim.Adam(self.network.parameters(), lr=learning_rate)
        
        # Hyperparameters
        self.gamma = gamma
        self.epsilon = epsilon_start
        self.epsilon_end = epsilon_end
        self.epsilon_decay = epsilon_decay
        self.batch_size = batch_size
        
        # Replay memory
        self.memory = deque(maxlen=memory_size)
        
    def _calculate_state_size(self) -> int:
        """Calculate flattened state size."""
        clo_plo_size = self.env.num_clos * self.env.num_plos
        assessment_size = self.env.num_clos * self.env.num_assessment_methods
        scores_size = self.env.num_clos
        syllabus_size = self.env.num_clos
        return clo_plo_size + assessment_size + scores_size + syllabus_size
        
    def _calculate_action_size(self) -> int:
        """Calculate total number of possible actions."""
        return self.env.num_clos * 4 * 2  # clos * issue_types * flag_options
        
    def _state_to_tensor(self, state: Dict) -> torch.Tensor:
        """Convert state dict to tensor."""
        state_list = []
        state_list.extend(state['clo_plo_mapping'].flatten())
        state_list.extend(state['assessment_methods'].flatten())
        state_list.extend(state['scores_valid'])
        state_list.extend(state['syllabus_match'])
        return torch.FloatTensor(state_list)
        
    def _action_to_dict(self, action_idx: int) -> Dict:
        """Convert action index to action dict."""
        clo_idx = action_idx // (4 * 2)
        remaining = action_idx % (4 * 2)
        issue_type = remaining // 2
        flag = remaining % 2
        return {
            'clo_index': clo_idx,
            'issue_type': issue_type,
            'flag': flag
        }
        
    def _dict_to_action(self, action_dict: Dict) -> int:
        """Convert action dict to action index."""
        return (action_dict['clo_index'] * 4 * 2 + 
                action_dict['issue_type'] * 2 + 
                action_dict['flag'])
        
    def select_action(self, state: Dict) -> Dict:
        """Select action using epsilon-greedy policy."""
        if random.random() < self.epsilon:
            action_idx = random.randrange(self.action_size)
            return self._action_to_dict(action_idx)
            
        with torch.no_grad():
            state_tensor = self._state_to_tensor(state)
            q_values = self.network(state_tensor)
            action_idx = q_values.argmax().item()
            return self._action_to_dict(action_idx)
            
    def train(self, num_episodes: int = 1000):
        """Train the agent."""
        for episode in range(num_episodes):
            state, _ = self.env.reset()
            episode_reward = 0
            
            while True:
                # Select and perform action
                action = self.select_action(state)
                next_state, reward, done, _, _ = self.env.step(action)
                episode_reward += reward
                
                # Store transition
                self.memory.append((state, action, reward, next_state, done))
                
                # Move to next state
                state = next_state
                
                # Perform one step of optimization
                self._optimize_model()
                
                if done:
                    break
            
            # Update target network
            if episode % 10 == 0:
                self.target_network.load_state_dict(self.network.state_dict())
            
            # Decay epsilon
            self.epsilon = max(self.epsilon_end, self.epsilon * self.epsilon_decay)
            
            print(f"Episode {episode + 1}/{num_episodes}, Reward: {episode_reward}, Epsilon: {self.epsilon:.2f}")
            
    def _optimize_model(self):
        """Perform one step of optimization."""
        if len(self.memory) < self.batch_size:
            return
            
        # Sample batch
        batch = random.sample(self.memory, self.batch_size)
        state_batch, action_batch, reward_batch, next_state_batch, done_batch = zip(*batch)
        
        # Convert to tensors
        state_tensor = torch.stack([self._state_to_tensor(s) for s in state_batch])
        action_tensor = torch.tensor([self._dict_to_action(a) for a in action_batch])
        reward_tensor = torch.tensor(reward_batch, dtype=torch.float32)
        next_state_tensor = torch.stack([self._state_to_tensor(s) for s in next_state_batch])
        done_tensor = torch.tensor(done_batch, dtype=torch.float32)
        
        # Compute current Q values
        current_q_values = self.network(state_tensor).gather(1, action_tensor.unsqueeze(1))
        
        # Compute target Q values
        with torch.no_grad():
            next_q_values = self.target_network(next_state_tensor).max(1)[0]
            target_q_values = reward_tensor + self.gamma * next_q_values * (1 - done_tensor)
            
        # Compute loss and optimize
        loss = nn.MSELoss()(current_q_values.squeeze(), target_q_values)
        self.optimizer.zero_grad()
        loss.backward()
        self.optimizer.step()
```

3. Create example usage in `examples/consistency_checker_example.py`:

```python
from agents.consistency_checker_agent import ConsistencyCheckerAgent
from agents.environments.consistency_checker_env import ConsistencyCheckEnvironment

def main():
    # Example documents
    documents = {
        'course_reports': [...],  # Your course reports
        'clo_assessments': [...],  # Your CLO assessments
        'syllabus': [...],  # Your syllabus data
    }
    
    # Create environment and agent
    env = ConsistencyCheckEnvironment(documents)
    agent = ConsistencyCheckerAgent(env)
    
    # Train agent
    agent.train(num_episodes=1000)
    
    # Use trained agent for consistency checking
    state, _ = env.reset()
    inconsistencies = []
    
    while True:
        action = agent.select_action(state)
        next_state, reward, done, _, _ = env.step(action)
        
        if action['flag'] == 1:
            inconsistencies.append({
                'clo_index': action['clo_index'],
                'issue_type': action['issue_type']
            })
            
        if done:
            break
        state = next_state
        
    print("Found inconsistencies:", inconsistencies)

if __name__ == "__main__":
    main()
```

### Implementation Details

1. **State Space**:
   - CLO-PLO mappings (binary matrix)
   - Assessment methods per CLO (binary matrix)
   - Score validity flags (binary vector)
   - Syllabus consistency flags (binary vector)

2. **Action Space**:
   - CLO index (which CLO to check)
   - Issue type (4 types: mapping, method, score, syllabus)
   - Flag (binary: issue/no issue)

3. **Reward Function**:
   - +1.0 for correctly identifying an inconsistency
   - +0.5 for correctly identifying consistency
   - -1.0 for incorrect identification

4. **Environment Features**:
   - Tracks actual inconsistencies
   - Provides detailed state information
   - Handles document processing
   - Manages episode termination

5. **Agent Features**:
   - Deep Q-Network based learning
   - Experience replay
   - Target network for stability
   - Epsilon-greedy exploration
   - Automatic epsilon decay

### Usage Instructions

1. Initialize the environment with your documents:
```python
env = ConsistencyCheckEnvironment(your_documents)
```

2. Create and train the agent:
```python
agent = ConsistencyCheckerAgent(env)
agent.train(num_episodes=1000)
```

3. Use the trained agent:
```python
state, _ = env.reset()
while True:
    action = agent.select_action(state)
    next_state, reward, done, _, _ = env.step(action)
    if done:
        break
    state = next_state
```

### Next Steps

1. [ ] Implement document processing functions
2. [ ] Add logging and monitoring
3. [ ] Implement save/load functionality
4. [ ] Add visualization tools
5. [ ] Create comprehensive test suite
